Introduce multiple storage drivers

GrzegorzDrozd opened the pull request:
Changes required by: https://github.com/perftools/xhgui/pull/273 (formerly https://github.com/perftools/xhgui/pull/268)

Introduce multiple storage drivers, currently supported are:
- pdo.sqlite (other drivers need testing)
- mongo
- files

Commit:
Additional options for profiler collector.

Add support for different serialization options: json, serialize,
igbinary.
Add support to save meta information to separate file - to speed up gui
operations.
Add information how to make output compatible with original gui and
other tools.
Add option to store additional data with profile information such as db
queries (copied from globals).
Fallback to default if config option is missing.

Commit:
test config

Commit:
Test fixes

Commit:
test fixes

Commit:
Database support

Commit:
PDO support
code clean up
config clean up

Commit:
store addtional data in profiles
remove some warnings/notices

markstory commented on composer.json:
```diff
@@ -1,5 +1,5 @@
 {
-    "name": "perftools/xhgui-collector",
+    "name": "GrzegorzDrozd/xhgui-collector",
```
This shouldn't change.

GrzegorzDrozd replied:
Yes, I know. I will revert it later when I am done with testing - need that to install package into application that I use for testing.

markstory commented on external/header.php:
```diff
@@ -161,30 +184,43 @@ function () {
+            'branch'            => !empty($GLOBALS['xhgui']['branch'])      ?: null,
+            'controller'        => !empty($GLOBALS['xhgui']['controller'])  ?: null,
+            'action'            => !empty($GLOBALS['xhgui']['action'])      ?: null,
+            'cookies'           => $_COOKIE,
```
Space aligning is a tedious process that causes diffs to be bigger than they need to be. Could this be restored to what it was before?

GrzegorzDrozd replied:
I am focusing right now on making sure all tests are fixed and new code is covered with tests.

markstory commented on src/Xhgui/Util.php:
```diff
@@ -20,4 +23,137 @@ public static function simpleUrl($url)
+            case 'json':
+                return json_encode($data);
+                break;
+            case 'serialize':
```
I don't think we should be using serialized at all. It is far too risky to have build an application that deserializes data obtained from the network.

GrzegorzDrozd replied:
As far as I remember original FB code uses serialization instead of json. Having this makes GUI compatible with FB code. I can remove that after I am done with tests and move it to separate PR for discussion.

markstory replied:
This project has never aimed to be compatible with the original Facebook data though.

Commit:
enable default file handler so that tests and new installations are working

Commit:
Fix: undefined key error

Commit:
Package name - lower case for composer 2.0

Commit:
Make sure that pdo saver uses existing id (in case of of import).

Commit:
Fix method detection
Fix id from mongo (with underscore)

Commit:
Fix query order

Commit:
Add ability to create object from instance of the factory - useful for testing

Commit:
Add filename generation to file saver.

Commit:
Provide path as a argument

Commit:
CR fix: remove facebook lib compatibility

Commit:
Merge upstream master

GrzegorzDrozd commented:
@markstory Ready for merge. But if I change package names from grzegorzdrozd/xhgui|grzegorzdrozd/xhgui-collector it will not build - unit tests will fail. Please review and when you are happy I will change them back to default, ok?

Commit:
Optimize saving: use 2 variables (path and filename that will default to
standard pattern)

Commit:
Sync default configs

Commit:
Add simple url as a separate field

glensc commented:
@GrzegorzDrozd could you please rebase to squash commits to a minimum. I can do the review as well.

glensc commented on composer.json:
```diff
@@ -23,7 +23,8 @@
         "ext-mongodb": "Mongo is needed to store profiler results for PHP > 7.",
         "ext-curl": "You need to install the curl extension to upload saved files",
-        "ext-pdo": "You need to install the pdo extension to store profiler results to a relational database",
+        "ext-json": "You ned json extension to work with json on some linux systems",
```
`ned` typo

glensc commented on external/header.php:
```diff
@@ -84,6 +87,11 @@
 }
 
+if (!extension_loaded('PDO') && Xhgui_Config::read('save.handler') === 'pdo') {
+    error_log('xhgui - extension pdo not loaded');
```
the handler should not produce noise on misconfiguration. either remove error log or throw a fatal exception.

GrzegorzDrozd replied:
What about original code above? 
`error_log('xhgui - extension mongo not loaded');`
Should I change that also ?

glensc commented on external/header.php:
```diff
@@ -95,30 +103,45 @@
 
+        // store session id before we close it:
+        $sessionId = null;
+        if (session_status() === PHP_SESSION_ACTIVE) {
```
this is not compatible with php 5.3 that project still targets

glensc commented on src/Xhgui/Saver.php:
```diff
@@ -30,21 +36,41 @@ public static function factory($config)
-                        $config['pdo']['pass']
-                    ),
-                    $config['pdo']['table']
+                return new Xhgui_Saver_PDO(
```
revert case change of the class

glensc commented on src/Xhgui/Saver.php:
```diff
@@ -30,21 +36,41 @@ public static function factory($config)
+                $mongo = new MongoClient(
+                    $config['db.host'],
+                    $config['db.options'] +
+                    [
```
short array syntax not compatible with still supported php 5.3

glensc commented on src/Xhgui/Saver/File.php:
```diff
@@ -1,17 +1,95 @@
+        $prefix = 'xhgui.data.'.microtime(true);
+        
+        if (empty($_SERVER['REQUEST_URI'])) {
+            if (version_compare(PHP_VERSION, '7.0.0') >= 0) {
```
use `PHP_VERSION_ID` constant instead of `version_compare` function

GrzegorzDrozd replied:
What do  you mean? PHP_VERSION_ID is available as of PHP 5.2.7, version_compare is available event before that.

glensc replied:
5.3.0 is the version being supported: https://github.com/perftools/xhgui-collector/blob/1.7.0/composer.json#L17

glensc commented on src/Xhgui/Saver/File.php:
```diff
@@ -1,17 +1,95 @@
+        if (empty($_SERVER['REQUEST_URI'])) {
+            if (version_compare(PHP_VERSION, '7.0.0') >= 0) {
+                try {
+                    $fileNamePattern = $prefix.bin2hex(random_bytes(5));
```
not sure about introducing such complex logic to this project. there exists [paragonie/random_compat](https://github.com/paragonie/random_compat) for this.

is such new dependency ok @markstory?

is such high randomness really needed @GrzegorzDrozd?

GrzegorzDrozd replied:
This would introduce another dependency that is php version dependant - there is no way to tell composer to install that package only for version lower than 7.
As for complexity - I am using xhgui in high traffic environment and I had few times where generated id was duplicated.

glensc replied:
yes, there is a way to tell composer exactly that. see random compat readme.

markstory replied:
If duplicates are a problem wouldn't a longer value give you the additional entropy to ensure uniqueness without taking on an additional dependency?

People still on PHP5 could just have a less production ready file saver as well. Php5 has long been EOL and there is a good case for dropping support entirely at this point.

glensc replied:
@markstory are you ok adding `paragonie/random_compat` dependency? I'll add it as separate PR, so this branch can just use `random_*` family functions.

ps: I'm not fine dropping PHP 5.3 myself, I still need it, and it's not maintenance burden to keep that support.

glensc replied:
@markstory wait. `paragonie/random_compat` is already here:  #11

@GrzegorzDrozd pelase use `random_*` family functions

GrzegorzDrozd replied:
I will change that to use random_*. Not sure how I missed it when I was syncing that with upstream.

Commit:
Code review fixes.

- revert short array syntax
- fix typos
- rename PDO class to Pdo
- change session existence check

Commit:
Code review fixes

- fix version compare

glensc commented on config/config.default.php:
```diff
@@ -1,53 +1,66 @@
-    'debug' => false,
-    'mode' => 'development',
+    'debug'     => false,
+    'mode'      => 'development',
```
please revert all formatting changes that do not change code.

glensc replied:
or if you insist formatting changes, we should do that in separate PR with a tool that controls it (php-cs-fixer for example).

markstory replied:
I too would prefer that we don't change the formatting of files.

glensc commented on config/config.default.php:
```diff
@@ -1,53 +1,66 @@
-    'date.format' => 'M jS H:i:s',
-    'detail.count' => 6,
-    'page.limit' => 25,
+    'run.view.filter.names' => array(
```
`run.view.filter.names` does not seem to exist. please remove and consider submitting as separate PR.

GrzegorzDrozd replied:
I wanted to keep that file in sync with xhgui config file. I will remove this.

Commit:
Code review changes

- remove unused config values
- remove unnecessary complex logic in filename generation

Commit:
Small fix - unnecessary variable.

Commit:
Code review changes

- remove unnecessary code formatting
- make sure no array is using short syntax

Commit:
Small fix: consistency.

glensc commented on src/Xhgui/Util.php:
```diff
@@ -21,17 +24,102 @@ public static function simpleUrl($url)
+        }
+
+        // try php 7+ function.
+        if (function_exists('random_bytes')) {
```
remove the `function_exists` check. and all the compatibility layer done below. the function is provided by php engine or the random-compat shim.

I wrote about this elsewhere in this PR.

glensc committed:
Merge upstream/master

glensc commented:
What is this different from https://github.com/perftools/xhgui-collector/pull/11? as it already added a PDO adapter?

glensc commented:
this is not suitable for merge or review. submit changes in smaller chunks and per feature.

currently, it seems to branched off before https://github.com/perftools/xhgui-collector/pull/11 was submitted, and impossible at the current state.
